feat(enhancement): Situation based Ship Animations Refactored

XessWaffle opened the pull request:
-----------------------
**Feature:** This PR implements the feature request detailed and discussed in issue #4991 

See Previous PRs for discussion: 
1. https://github.com/endless-sky/endless-sky/pull/7150
2. https://github.com/endless-sky/endless-sky/pull/8797

## Feature Details
See Wiki PR for updated details: https://github.com/endless-sky/endless-sky-wiki/pull/31

To rehash the issue, the goal is to have a ship play different animations based on the situation it is in (eg. flying, landing etc.).

This PR currently implements six different states: Flying, Fighting, Landing, Launching, Jumping, Disabled.
- Flying: Implies that the ship is simply moving through space. This is also the default sprite i.e. if another sprite is not defined, the sprite that will be displayed is the flying sprite. For example, if a ship is defined without a landing, fighting or launching sprite, then the flying sprite will be displayed in all of those cases. Furthermore, bodies that are defined with just the "sprite" token will be loaded into the flying sprite. 
- Firing (changed from Fighting): Implies that the ship has an active enemy target i.e. the target is not disabled and it is in the system and in weapons range (thanks @Hurleveur for the weaponRange suggestion), or is targeting an asteroid. The body may or may not be firing its weapons in this state. The players ship also enters this state if their primary weapon is fired, regardless of them having a target.
- Landing: Implies that the ship is currently descending to land on a planet or wormhole.
- Launching: Implies that the ship is currently ascending from a planet or wormhole.
- Jumping: Implies that the ship is currently trying to move through hyperspace (before a jump) or is moving through hyperspace (during a jump).
- Disabled: Implies that the ship is currently disabled.

To Implement:
- [x] Firing: Generic firing animation for a ship (uncovering hardpoints etc.) (@Zitchas)
- [x] Special firing animations: Certain weapons on certain ships should trigger special animations (@Zitchas) 
- [x] "delay" keyword: A delay before state change (@Azure3141)
- [x] Disabled State: Should transition the ship out of the current state (state -> disabled) and also account for being rescued (disabled -> flying) (@Azure3141).
- [x] "random": If this keyword appears in a state, pick frames randomly while in the state (@Saugia)
- [x] "indicate" update: Allow indicate to have a value from zero to one identifying the percentage of the animation that must complete before the specified action is allowed (@Ferociousfeind).
- [x] Special animations update: Instead of triggering an animated sprite if the ship has the outfit, trigger the sprite if the ship is using the outfit.
- [x] Landing and launching for fighter style ships. Launching would imply deploying fighters, and landing would imply recalling them.

Potential future PRs:
- Hyperspacing substate: Should only occur if the ship is in the Jumping state, and the ship is travelling through hyperspace, will be implemented alongside special firing animation as a special jumping animation (e.g. play a different animation for a jump drive etc.) This is now possible as a trigger based animation.
- Thrusting/Overheating visuals: May not be as efficient to implement as states, but may be possible to implement via shaders. (@Azure3141). Might make this a separate PR.


The PR also handles transitions between different states. For example, if transitioning from launching to flying, completing the launching animation before going into the flying animation would be something that is preferred in some cases. A couple of new animation keywords have been added:
- "transition rewind": If a state change is requested (e.g. flying -> fighting), play the sprite animation associated with the current state (not the requested state) backwards until it reaches the first frame, then transition to the requested state.
- "transition finish": If a state change is requested, finish the current loop of the animation and transition states.
If neither keyword is used, then the transition between sprite animations is immediate. 
- "transition delay": Once a state change is requested, wait a certain number of frames, then complete the transition. Can be used to ensure that states aren't changing too rapidly. During the delay period, the transition may be cancelled as well. (@Azure3141 thanks for the suggestion!)
- "indicate": Typically used during the Fighting or Jumping states, essentially, this keyword will make the animation non-repeatable and indicate to the ship when the last frame is reached. Upon reaching the last frame, the ship can perform the desired action (firing its weapons or jumping). For example, if "indicate" is used in "sprite-firing", then the ship will not fire until the animation is complete.
- "indicate percentage [value]": Similar to the indicate state, except only a certain percentage of the animation has to finish before the action can be performed.
- "random": Used when a sprite should pick random frames in a certain state. In the example below, "random" is used in the disabled state to pick frames from the flying animation randomly. (@Saugia thanks for the suggestion!)
- "ramp [ramp in rate] [ramp out rate]": Used when an animation frame rate should ramp up or down. Ramp in rate represents the rate at which the frameRate changes when the body has just transitioned into a state, and ramp out is the rate at which the frameRate changes when the body is transitioning into another state. The ramp rates can be both negative and positive (or zero).
- "reverse": Used to simply reverse the frame number displayed. For example, in an animation with 20 frames, if we are currently on frame 15, an animation with the reverse keyword will be on frame 5. Essentially, the animation will be played backward.


The last keyword is somewhat more complicated: "trigger [path-to-animated-sprite]"
- Essentially, it is a conditional sprite that can be played when the ship is in certain states. For example, if a ship is in the firing state, and we want it to play a different animation if a certain weapon is firing, a trigger can be setup for that situation.
- Several different preliminary conditions have been implemented:
  - "hull": Gives the amount of hull currently on the ship.
  - "shields: Gives the amount of shields currently on the ship.
  - "outfit (installed): [outfit]": Returns the number of [outfit] installed on the ship.
  - "jump (using): [outfit]": Returns 1 if the current jump is happening with [outfit] where [outfit] can be any sort of drive. If [outfit] is not a drive, then returns 0.
  - "weapon (firing): [outfit]": Returns 1 if [outfit] is a weapon and is firing in the current frame.
- Trigger sprites can also be individually animated with the same parameters as listed above.

## UI Screenshots
See Closed PR https://github.com/endless-sky/endless-sky/pull/8797

## Usage Examples
I used these parameters with a ship I modeled called the Raptor. The plugin is linked below and I have attached a save file to this PR. Note that you can animate each different state's sprite individually, and as stated before, not all states need to have a transition defined. Furthermore, the "sprite" or "sprite-flying" keywords can be used interchangeably.
```
	sprite "ship/raptor/flying-disabled/raptor"
		"frame rate" 5
	sprite-firing "ship/raptor/fighting/raptor"
                ramp 20 0
		trigger ship/raptor/fighting/sunbeam/raptor
                         conditions
                                   has "outfit (installed): Sunbeam"
			"frame rate" 25
			"indicate percentage" 0.5
		"frame rate" 20
		"indicate percentage" 0.75
		"transition delay" 50
		"transition rewind"
	sprite-landing "ship/raptor/landing/raptor"
		"frame rate" 10
		"no repeat"
	sprite-launching "ship/raptor/launching/raptor"
		"frame rate" 5
		"no repeat"
		"transition finish"
	sprite-jumping "ship/raptor/jumping/raptor"
                trigger "Jump Drive" "ship/raptor/jumping/jump drive/raptor" "on use"
		"frame rate" 20
		"transition rewind"
	sprite-disabled "ship/raptor/flying-disabled/raptor"
		"frame rate" 5
		"random"
```
Plugin: https://github.com/XessWaffle/endless-sky-animation-test
- Helps to have the all content plugin installed as well!

## Testing Done
So far, I have only tested my ship with the parameters above, and ensured that other animations (e.g the Peregrine, or the Shuttle) were unchanged and uninterrupted by my changes.

Need to test

- [x] AI fleets that have ships with animations
- [x] Special ships with animations
- [x] Transitions between edge case states i.e. Firing -> Landing: Noticed a jumpy bug when rapidly switching between jumping, firing and flying
- [x] Jumping animations
- [x] Mining ships with animations (for firing on asteroids)
- [x] Can the game be played normally as is without being affected (all storylines, game mechanics etc.): Played through intro missions with Star barge without any noticeable sprite related bugs.

Use ESLauncher2 to run locally!
## Performance Impact
More dynamic memory is used to load in different sprites and animation parameters for a certain ship, could lead to performance impacts if animations get large, but otherwise the performance impact should be minimal.

RisingLeaf commented on source/SpriteParameters.h:
```diff
@@ -0,0 +1,99 @@
+
+class Sprite;
+
+// Class holding all of the animation parameters required to animate a sprite
```
```suggestion


// Class holding all of the animation parameters required to animate a sprite.
```

RisingLeaf commented on source/SpriteParameters.h:
```diff
@@ -0,0 +1,99 @@
+		bool transitionRewind = false;
+		bool indicateReady = false;
+	};
+
```
```suggestion


```

RisingLeaf commented on source/SpriteParameters.h:
```diff
@@ -0,0 +1,99 @@
+	// A node defining the parameters for a certain trigger sprite
+	class AnimationParameters {
+	public:
+		// Act like a struct
+		// Animation parameters as found in Body.h
+		float frameRate = 2. / 60.;
+		float rampUpRate = 0.0f;
+		float rampDownRate = 0.0f;
+		float startFrame = 0.f;
+		float scale = 1.f;
+		float indicatePercentage = -1.0f;
+		int delay = 0;
+		int transitionDelay = 0;
+
+		bool startAtZero = false;
+		bool randomize = false;
+		bool randomizeStart = false;
+		bool repeat = true;
+		bool rewind = false;
+		bool reverse = false;
+
+		// Defines what to do when a state transition is requested (eg. FLYING to LANDING)
+		bool transitionFinish = false;
+		bool transitionRewind = false;
+		bool indicateReady = false;
```
more comments on the variables would be nice

RisingLeaf commented on source/SpriteParameters.h:
```diff
@@ -0,0 +1,99 @@
+public:
+	typedef std::tuple<const Sprite*, SpriteParameters::AnimationParameters, ConditionSet> SpriteDetails;
+	typedef std::map<int, SpriteParameters::SpriteDetails> SpriteMap;
+
```
```suggestion


```

RisingLeaf commented on source/SpriteParameters.h:
```diff
@@ -0,0 +1,99 @@
+	SpriteParameters();
+	explicit SpriteParameters(const Sprite *sprite);
+	// Add a sprite-trigger mapping
+	void SetSprite(int index, const Sprite *sprite,
+				   SpriteParameters::AnimationParameters data, ConditionSet triggerConditions);
```
```suggestion
	void SetSprite(int index, const Sprite *sprite, SpriteParameters::AnimationParameters data, 
		ConditionSet triggerConditions);
```

RisingLeaf commented on source/SpriteParameters.h:
```diff
@@ -0,0 +1,99 @@
+public:
+	SpriteParameters();
+	explicit SpriteParameters(const Sprite *sprite);
+	// Add a sprite-trigger mapping
```
```suggestion
	// Add a sprite-trigger mapping.
```

RisingLeaf commented on source/SpriteParameters.h:
```diff
@@ -0,0 +1,99 @@
+	// Add a sprite-trigger mapping
+	void SetSprite(int index, const Sprite *sprite,
+				   SpriteParameters::AnimationParameters data, ConditionSet triggerConditions);
+	// Get the data associated with the current trigger
+	const Sprite *GetSprite(int index = -1) const;
+	ConditionSet GetConditions(int index = -1) const;
+	AnimationParameters GetParameters(int index = -1) const;
+	// Used to verify trigger transitions
+	const int GetExposedID() const;
+	// Check all conditions and return if one is true
+	const int RequestTriggerUpdate(ConditionsStore &store);
+	// Complete the switch to a new sprite
+	void CompleteTriggerRequest();
+	// Used for saving the sprites.
+	const SpriteMap *GetAllSprites() const;
```
```suggestion
	// Get the data associated with the current trigger.
	const Sprite *GetSprite(int index = -1) const;
	ConditionSet GetConditions(int index = -1) const;
	AnimationParameters GetParameters(int index = -1) const;
	// Used to verify trigger transitions.
	const int GetExposedID() const;
	// Check all conditions and return if one is true.
	const int RequestTriggerUpdate(ConditionsStore &store);
	// Complete the switch to a new sprite.
	void CompleteTriggerRequest();
	// Used for saving the sprites.
	const SpriteMap *GetAllSprites() const;
```

RisingLeaf commented on source/SpriteParameters.h:
```diff
@@ -0,0 +1,99 @@
+	const SpriteMap *GetAllSprites() const;
+
+
+public:
+	// ID For default sprite
+	const static int DEFAULT = 0;
+	// Animation parameters exposed to Body
+	SpriteParameters::AnimationParameters exposed;
+
```
the second variable should be private with a getter

RisingLeaf commented on source/SpriteParameters.cpp:
```diff
@@ -0,0 +1,151 @@
+#include "SpriteParameters.h"
+
+#include "Sprite.h"
+
```
```suggestion



```

RisingLeaf commented on source/SpriteParameters.cpp:
```diff
@@ -0,0 +1,151 @@
+SpriteParameters::SpriteParameters(const Sprite *sprite)
+{
+	AnimationParameters initDefault;
+	static ConditionSet empty;
+	auto tuple = SpriteParameters::SpriteDetails{sprite, initDefault, empty};
```
```suggestion
	auto tuple = SpriteParameters::SpriteDetails{sprite, initDefault, {}};
```
wouldnt this work too?

RisingLeaf commented on source/SpriteParameters.cpp:
```diff
@@ -0,0 +1,151 @@
+
+
+
+void SpriteParameters::SetSprite(int index, const Sprite *sprite,
+									SpriteParameters::AnimationParameters params, ConditionSet triggerConditions)
```
```suggestion
void SpriteParameters::SetSprite(int index, const Sprite *sprite, SpriteParameters::AnimationParameters params, 
	ConditionSet triggerConditions)
```

RisingLeaf commented on source/SpriteParameters.cpp:
```diff
@@ -0,0 +1,151 @@
+									SpriteParameters::AnimationParameters params, ConditionSet triggerConditions)
+{
+	auto tuple = SpriteParameters::SpriteDetails{sprite, params, triggerConditions};
+	this->sprites.insert(std::pair<int, SpriteParameters::SpriteDetails>(index, tuple));
+	if(index == DEFAULT)
+	{
+		this->exposed = params;
+		this->exposedIndex = 0;
+		this->defaultDetails = tuple;
+		this->exposedDetails = tuple;
+	}
```
why are there so many this-> I do not see why they are needed here

RisingLeaf commented on source/SpriteParameters.cpp:
```diff
@@ -0,0 +1,151 @@
+{
+	if(index < 0)
+		return std::get<0>(this->exposedDetails);
+	else
+	{
+		auto it = this->sprites.find(index);
+		if(it != this->sprites.end())
+			return std::get<0>(it->second);
+	}
```
```suggestion

	auto it = this->sprites.find(index);
	if(it != this->sprites.end())
		return std::get<0>(it->second);

```

RisingLeaf commented on source/SpriteParameters.cpp:
```diff
@@ -0,0 +1,151 @@
+	static ConditionSet empty;
+	if(index < 0)
+		return std::get<2>(this->exposedDetails);
+	else
+	{
+		auto it = this->sprites.find(index);
+		if(it != this->sprites.end())
+			return std::get<2>(it->second);
+	}
```
```suggestion

	auto it = this->sprites.find(index);
	if(it != this->sprites.end())
		return std::get<2>(it->second);

```

RisingLeaf commented on source/SpriteParameters.cpp:
```diff
@@ -0,0 +1,151 @@
+	SpriteParameters::AnimationParameters def;
+	if(index < 0)
+		return std::get<1>(this->exposedDetails);
+	else
+	{
+		auto it = this->sprites.find(index);
+		if(it != this->sprites.end())
+			return std::get<1>(it->second);
+	}
```
```suggestion

	auto it = this->sprites.find(index);
	if(it != this->sprites.end())
		return std::get<1>(it->second);

```

RisingLeaf commented on source/SpriteParameters.cpp:
```diff
@@ -0,0 +1,151 @@
+			}
+
+		}
+	// Return to default
```
```suggestion

	// Return to default.
```

RisingLeaf commented on source/SpriteParameters.cpp:
```diff
@@ -0,0 +1,151 @@
+
+const int SpriteParameters::RequestTriggerUpdate(ConditionsStore &store)
+{
+	for(auto it = this->sprites.begin(); it != this->sprites.end(); it++)
+		if(it->first != DEFAULT)
+		{
+			ConditionSet conditions = std::get<2>(it->second);
+			if(conditions.Test(store))
+			{
+				this->requestedIndex = it->first;
+				return it->first;
+			}
+
+		}
```
```suggestion
	for(auto it : this->sprites)
		if(it->first != DEFAULT)
		{
			ConditionSet conditions = std::get<2>(it->second);
			if(conditions.Test(store))
			{
				this->requestedIndex = it->first;
				return it->first;
			}

		}
```

RisingLeaf commented on source/Ship.cpp:
```diff
@@ -4788,69 +4896,112 @@ void Ship::DoMovement(bool &isUsingAfterburner)
 
 void Ship::StepTargeting()
 {
+	static const double weaponsRangeMultiplier = 1.25;
```
```suggestion
	static const double WEAPONS_RANGE_MULTIPLIER= 1.25;
```

RisingLeaf commented on source/Body.h:
```diff
@@ -33,6 +36,11 @@ class Sprite;
+	// States for animation purposes
+	enum BodyState{FLYING, FIRING, LAUNCHING, LANDING, JUMPING, DISABLED, NUM_STATES, CURRENT, TRIGGER};
+	// Minimum/Default frame rate of animations
+	static constexpr float MIN_FRAME_RATE = 2. / 60.;
```
```suggestion
	static constexpr float MIN_FRAME_RATE = 2. / 60.;


```

RisingLeaf reviewed (commented):
Just style for now, please remove all unnecessary this->

Maryn-s commented:
Very cool! Another animation that could be thrown in there would be something like "deploying" when the ship launches its fighters/drones. I can also see this being expanded to allow for hardpoint repositioning under certain animations, like guns facing a different direction or thursters and steering engines being moved around to behave as the other, though due to the complications that could be its own PR.

beccabunny commented:
I haven't seen it mentioned in the OP - when fighters are docked to a carrier, what sprite do they show? Is it just the normal sprite, and/or is there a way to make them use the first frame of the "launching" state (or the last frame of "landing") with the available keywords?

XessWaffle commented:
Hi all, thanks for taking a look at this, I (hopefully) fixed all of the suggested style changes/code organization!

> Very cool! Another animation that could be thrown in there would be something like "deploying" when the ship launches its fighters/drones, like opening hanger doors or shifting its wings to allow for a clear launch. I can also see this being expanded to allow for hardpoint repositioning under certain animations, like guns facing a different direction or thursters and steering engines being moved around to behave as the other, though due to the complications that could be its own PR.

Having a deploy animation is an interesting idea. I would have to look into whether having just a deploy state would suffice or both deploy + recall states would be necessary in order to make this work. As for the hardpoint motion, this would fall under its own PR as it would depend a lot on the changes made through this PR.


> I haven't seen it mentioned in the OP - when fighters are docked to a carrier, what sprite do they show? Is it just the normal sprite, and/or is there a way to make them use the first frame of the "launching" state (or the last frame of "landing") with the available keywords?

When they are docked to the carrier, they will play the sprite animation defined for the LANDING state. If you have the "no repeat" flag defined on the LANDING state, then after it reaches the last sprite, it will stay there until redeployed. Once its deployed it will automatically go to the first sprite in the LAUNCHING state. Note, there is no code which checks that the first played sprite in the LAUNCHING state is the same as the last played sprite in the LANDING state so the animator would have to ensure that the sprites used would create a seamless experience.

Hyugat commented:
I love this. I'd very much welcome the ability to have deploy animations and separate flying/docked fighter sprites. I'm imagining something similar to how an MV-22 Osprey can fold up its wings to pack into a tighter space. That would allow for a lot more variation in fighter design, and this pr as a whole I think really opens up what people can do. Variable geometry spacecraft are absolutely something I'd love to see in this game and this is a huge step towards making that happen.

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -20,6 +20,7 @@ this program. If not, see <https://www.gnu.org/licenses/>.
 #include "GameData.h"
 #include "Mask.h"
 #include "MaskManager.h"
+#include "Outfit.h"
```
Where is this used?

XessWaffle replied:
Good catch, I had initially used this to check whether certain sprites should be triggered if some outfits were installed but now I use the ConditionStore. Will remove this include!

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -34,8 +35,14 @@ using namespace std;
-	: position(position), velocity(velocity), angle(facing), zoom(zoom), sprite(sprite), randomize(true)
+	: position(position), velocity(velocity), angle(facing), zoom(zoom)
 {
+	SpriteParameters *spriteState = &sprites[Body::BodyState::FLYING];
```
```suggestion
	SpriteParameters *spriteState = &sprites[BodyState::FLYING];
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -34,8 +35,14 @@ using namespace std;
+	SpriteParameters *spriteState = &sprites[Body::BodyState::FLYING];
+	SpriteParameters::AnimationParameters spriteAnimationParameters;
+	spriteAnimationParameters.randomizeStart = true;
+	static ConditionSet empty;
+	spriteState->SetSprite(SpriteParameters::DEFAULT, sprite, spriteAnimationParameters, empty);
```
```suggestion
	spriteState->SetSprite(SpriteParameters::DEFAULT, sprite, spriteAnimationParameters, {});
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -55,22 +62,50 @@ Body::Body(const Body &sprite, Point position, Point velocity, Angle facing, dou
+
+
+// Check that this Body has a sprite for a specific BodyState.
+bool Body::HasSpriteFor(Body::BodyState state) const
```
```suggestion
bool Body::HasSpriteFor(BodyState state) const
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -55,22 +62,50 @@ Body::Body(const Body &sprite, Point position, Point velocity, Angle facing, dou
+const Sprite *Body::GetSprite(Body::BodyState state) const
 {
-	return sprite;
+	Body::BodyState selected = state != Body::BodyState::CURRENT ? state : currentState;
```
```suggestion
	BodyState selected = state != BodyState::CURRENT ? state : currentState;
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -55,22 +62,50 @@ Body::Body(const Body &sprite, Point position, Point velocity, Angle facing, dou
+
+	// Return flying sprite if the requested state's sprite does not exist.
+	const Sprite *sprite = spriteState->GetSprite();
+	if(sprite != nullptr && !returnDefaultSprite)
```
```suggestion
	if(sprite && !returnDefaultSprite)
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -55,22 +62,50 @@ Body::Body(const Body &sprite, Point position, Point velocity, Angle facing, dou
+	if(sprite != nullptr && !returnDefaultSprite)
+		return sprite;
+	else
+		return sprites[Body::BodyState::FLYING].GetSprite();
```
```suggestion
		return sprites[BodyState::FLYING].GetSprite();
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -55,22 +62,50 @@ Body::Body(const Body &sprite, Point position, Point velocity, Angle facing, dou
+
+
+
+Body::BodyState Body::GetState() const
```
```suggestion
BodyState Body::GetState() const
```

XessWaffle replied:
Hitting a compilation error when making this change for some reason.

TomGoodIdea replied:
ah yes sorry, forgot that here you are technically outside the class

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -104,10 +140,21 @@ int Body::GetSwizzle() const
-	return frame;
+		finishStateTransition = SetStep(step);
+	// Choose between frame, random frame and reversed frame, based on the corresponding parameters.
+	float returnFrame = 0.0;
```
style
```suggestion
	float returnFrame = 0.;
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -205,90 +251,399 @@ const Government *Body::GetGovernment() const
+		return false;
+
+	// Static map for loading sprites.
+	static const std::map<const std::string, Body::BodyState> SPRITE_LOAD_PARAMS = {
+		{"sprite", Body::BodyState::FLYING},
+		{"sprite-flying", Body::BodyState::FLYING},
+		{"sprite-firing", Body::BodyState::FIRING},
+		{"sprite-launching", Body::BodyState::LAUNCHING},
+		{"sprite-landing", Body::BodyState::LANDING},
+		{"sprite-jumping", Body::BodyState::JUMPING},
+		{"sprite-disabled", Body::BodyState::DISABLED},
+		{"hardpoint sprite", Body::BodyState::FLYING},
+		{"flare sprite", Body::BodyState::FLYING},
+		{"reverse flare sprite", Body::BodyState::FLYING},
+		{"steering flare sprite", Body::BodyState::FLYING}
```
```suggestion
	static const map<const string, BodyState> SPRITE_LOAD_PARAMS = {
		{"sprite", BodyState::FLYING},
		{"sprite-flying", BodyState::FLYING},
		{"sprite-firing", BodyState::FIRING},
		{"sprite-launching", BodyState::LAUNCHING},
		{"sprite-landing", BodyState::LANDING},
		{"sprite-jumping", BodyState::JUMPING},
		{"sprite-disabled", BodyState::DISABLED},
		{"hardpoint sprite", BodyState::FLYING},
		{"flare sprite", BodyState::FLYING},
		{"reverse flare sprite", BodyState::FLYING},
		{"steering flare sprite", BodyState::FLYING}
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -205,90 +251,399 @@ const Government *Body::GetGovernment() const
+	};
+
+	// Default to flying state sprite.
+	Body::BodyState state = Body::BodyState::FLYING;
```
```suggestion
	BodyState state = BodyState::FLYING;
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -205,90 +251,399 @@ const Government *Body::GetGovernment() const
+
+	// Default to flying state sprite.
+	Body::BodyState state = Body::BodyState::FLYING;
+	auto pair = SPRITE_LOAD_PARAMS.find(node.Token(0));
```
I'd avoid naming a variable "pair", because at some point some compiler may start to dislike it.

XessWaffle replied:
Makes sense, changed it to `flyingPair`

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -205,90 +251,399 @@ const Government *Body::GetGovernment() const
+	const Sprite *sprite = SpriteSet::Get(node.Token(1));
+	SpriteParameters *spriteData = &sprites[state];
+	SpriteParameters::AnimationParameters spriteAnimationParameters;
+	std::vector<DataNode> triggerSpriteDefer;
```
```suggestion
	vector<DataNode> triggerSpriteDefer;
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -205,90 +251,399 @@ const Government *Body::GetGovernment() const
+				spriteAnimationParameters.rampDownRate = child.Value(2) / 3600.;
+			}
+		}
+		else if(child.Token(0) == "indicate percentage" && child.Size() >= 2 && child.Value(1) > 0. && child.Value(1) < 1.)
```
Percentage is a value between 0 and 100. Internally you can store it as a [0,1] float as long as it is well documented, but if you have an attribute called percentage, it either should not be called as such, or should require a real percentage.

XessWaffle replied:
Agree with this, I will go ahead and divide the value by `100.` therefore expecting an actual percent from the sprite definition instead of a float.

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -205,90 +251,399 @@ const Government *Body::GetGovernment() const
+			spriteAnimationParameters.indicateReady = true;
+			spriteAnimationParameters.repeat = false;
+			spriteAnimationParameters.startAtZero = true;
+			spriteAnimationParameters.indicatePercentage = 1.0f;
```
```suggestion
			spriteAnimationParameters.indicatePercentage = 1.f;
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -205,90 +251,399 @@ const Government *Body::GetGovernment() const
 	}
 
+	// Now load the trigger sprites.
+	for(int val = 0; val < static_cast<int>(triggerSpriteDefer.size()); ++val)
```
```suggestion
	for(size_t val = 0; val < triggerSpriteDefer.size(); ++val)
```
also why "val" and not just "i"?

XessWaffle replied:
Not sure, maybe I was feeling bored of i when writing this part xD

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -205,90 +251,399 @@ const Government *Body::GetGovernment() const
+	// Now load the trigger sprites.
+	for(int val = 0; val < static_cast<int>(triggerSpriteDefer.size()); ++val)
+	{
+		DataNode node = triggerSpriteDefer.at(val);
```
You don't need bounds checking here as the loop is limited to the size of the vector.
```suggestion
		DataNode node = triggerSpriteDefer[val];
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -205,90 +251,399 @@ const Government *Body::GetGovernment() const
+		DataNode node = triggerSpriteDefer.at(val);
+		LoadTriggerSprite(node, state, spriteAnimationParameters, val + 1);
+	}
+	static ConditionSet empty;
+	spriteData->SetSprite(SpriteParameters::DEFAULT, sprite, spriteAnimationParameters, empty);
```
```suggestion
	spriteData->SetSprite(SpriteParameters::DEFAULT, sprite, spriteAnimationParameters, {});
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -205,90 +251,399 @@ const Government *Body::GetGovernment() const
-void Body::SaveSprite(DataWriter &out, const string &tag) const
+// Returns whether the trigger sprite is based on the outfit being used.
+void Body::LoadTriggerSprite(const DataNode &node, Body::BodyState state,
+							SpriteParameters::AnimationParameters params, int index)
```
```suggestion
	SpriteParameters::AnimationParameters params, int index)
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -205,90 +251,399 @@ const Government *Body::GetGovernment() const
+// Save the sprite specification, including all animation attributes.
+void Body::SaveSprite(DataWriter &out, const string &tag, bool allStates) const
+{
+	const std::string tags[Body::BodyState::NUM_STATES] = {allStates ? "sprite" : tag, "sprite-firing",
```
```suggestion
	const string tags[BodyState::NUM_STATES] = {allStates ? "sprite" : tag, "sprite-firing",
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -205,90 +251,399 @@ const Government *Body::GetGovernment() const
+void Body::SaveSprite(DataWriter &out, const string &tag, bool allStates) const
+{
+	const std::string tags[Body::BodyState::NUM_STATES] = {allStates ? "sprite" : tag, "sprite-firing",
+												"sprite-launching", "sprite-landing",
+												"sprite-jumping", "sprite-disabled"};
```
```suggestion
		"sprite-launching", "sprite-landing",
		"sprite-jumping", "sprite-disabled"};
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -205,90 +251,399 @@ const Government *Body::GetGovernment() const
+												"sprite-launching", "sprite-landing",
+												"sprite-jumping", "sprite-disabled"};
+
+	for(int i = 0; i < Body::BodyState::NUM_STATES; i++)
```
```suggestion
	for(int i = 0; i < BodyState::NUM_STATES; i++)
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -205,90 +251,399 @@ const Government *Body::GetGovernment() const
+			{
+				SaveSpriteParameters(out, spriteState, 0);
+				// The map of all spriteStates.
+				auto map = spriteState->GetAllSprites();
```
Possible "map" name ambiguity (see one of my previous comments).

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -205,90 +251,399 @@ const Government *Body::GetGovernment() const
+				// The map of all spriteStates.
+				auto map = spriteState->GetAllSprites();
+
+				for(auto it = map->begin(); it != map->end(); ++it)
+					if(it->first != 0)
+					{
+						const Sprite *triggerSprite = spriteState->GetSprite(it->first);
+						out.Write("trigger", triggerSprite->Name());
+						out.BeginChild();
+						{
+							out.Write("conditions");
+							out.BeginChild();
+							{
+								spriteState->GetConditions(it->first).Save(out);
+							}
+							out.EndChild();
+							SaveSpriteParameters(out, spriteState, it->first);
```
```suggestion
				for(auto &it : map)
					if(it.first)
					{
						const Sprite *triggerSprite = spriteState->GetSprite(it.first);
						out.Write("trigger", triggerSprite->Name());
						out.BeginChild();
						{
							out.Write("conditions");
							out.BeginChild();
							{
								spriteState->GetConditions(it.first).Save(out);
							}
							out.EndChild();
							SaveSpriteParameters(out, spriteState, it.first);
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -205,90 +251,399 @@ const Government *Body::GetGovernment() const
+void Body::SaveSpriteParameters(DataWriter &out, SpriteParameters *state, int index) const
+{
+	SpriteParameters::AnimationParameters exposed = state->GetParameters(index);
+	if(exposed.frameRate != static_cast<float>(Body::MIN_FRAME_RATE))
```
```suggestion
	if(exposed.frameRate != static_cast<float>(MIN_FRAME_RATE))
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -205,90 +251,399 @@ const Government *Body::GetGovernment() const
+		out.Write("center", exposed.center.X(), exposed.center.Y());
+	if(exposed.reverse)
+		out.Write("reverse");
+	if(exposed.rampUpRate != 0.0f || exposed.rampDownRate != 0.0f)
```
```suggestion
	if(exposed.rampUpRate || exposed.rampDownRate)
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -205,90 +251,399 @@ const Government *Body::GetGovernment() const
+		out.Write("ramp", exposed.rampUpRate * 3600., exposed.rampDownRate * 3600.);
+	if(exposed.indicateReady)
+	{
+		if(exposed.indicatePercentage == 1.0f)
```
```suggestion
		if(exposed.indicatePercentage == 1.f)
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -205,90 +251,399 @@ const Government *Body::GetGovernment() const
 
 // Set the sprite.
-void Body::SetSprite(const Sprite *sprite)
+void Body::SetSprite(const Sprite *sprite, Body::BodyState state)
```
```suggestion
void Body::SetSprite(const Sprite *sprite, BodyState state)
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -205,90 +251,399 @@ const Government *Body::GetGovernment() const
 {
-	this->sprite = sprite;
+	SpriteParameters::AnimationParameters init;
+	static ConditionSet empty;
+	sprites[state].SetSprite(SpriteParameters::DEFAULT, sprite, init, empty);
```
```suggestion
	sprites[state].SetSprite(SpriteParameters::DEFAULT, sprite, init, {});
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -205,90 +251,399 @@ const Government *Body::GetGovernment() const
 
 
+// Set the state.
+void Body::SetState(Body::BodyState state)
```
```suggestion
void Body::SetState(BodyState state)
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -205,90 +251,399 @@ const Government *Body::GetGovernment() const
+	// If the transition has a delay, ensure that rapid changes from transitions
+	// which ignore the delay to transitions that don't (ignore delays)
+	// keep smooth animations.
+	bool delayIgnoredPreviousStep = transitionState == Body::BodyState::JUMPING ||
+									transitionState == Body::BodyState::DISABLED ||
+									transitionState == Body::BodyState::LANDING;
+	bool delayActiveCurrentStep = (state == Body::BodyState::FLYING ||
+									state == Body::BodyState::LAUNCHING || state == Body::BodyState::FIRING)
+									&& delayed < anim.transitionDelay;
+	if(state == currentState && transitionState != currentState
+			&& transitionState != Body::BodyState::TRIGGER)
```
```suggestion
	bool delayIgnoredPreviousStep = transitionState == BodyState::JUMPING ||
		transitionState == BodyState::DISABLED ||
		transitionState == BodyState::LANDING;
	bool delayActiveCurrentStep = (state == BodyState::FLYING ||
		state == BodyState::LAUNCHING || state == BodyState::FIRING)
		&& delayed < anim.transitionDelay;
	if(state == currentState && transitionState != currentState
		&& transitionState != BodyState::TRIGGER)
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -205,90 +251,399 @@ const Government *Body::GetGovernment() const
+		}
+	}
+
+	transitionState = transitionState != Body::BodyState::TRIGGER ? state : Body::BodyState::TRIGGER;
```
```suggestion
	transitionState = transitionState != BodyState::TRIGGER ? state : BodyState::TRIGGER;
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -328,6 +683,80 @@ void Body::PauseAnimation()
 
+void Body::ShowDefaultSprite(bool defaultSprite)
+{
+	this->returnDefaultSprite = defaultSprite;
```
```suggestion
	returnDefaultSprite = defaultSprite;
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -328,6 +683,80 @@ void Body::PauseAnimation()
+// E.g. if FIRING a certain weapon, check if we have an animation for that weapon, and trigger it.
+void Body::CheckTriggers()
+{
+	for(int i = 0; i < BodyState::NUM_STATES; i++)
```
Could we use a C++11 loop here?

XessWaffle replied:
Possible, but I would have to create an array of the elements of the enum as a const. If that is something that is preferred according to the code style, I am happy to make that change.

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -328,6 +683,80 @@ void Body::PauseAnimation()
+void Body::CheckTriggers()
+{
+	for(int i = 0; i < BodyState::NUM_STATES; i++)
+		if(HasSpriteFor(static_cast<Body::BodyState>(i)))
```
```suggestion
		if(HasSpriteFor(static_cast<BodyState>(i)))
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -328,6 +683,80 @@ void Body::PauseAnimation()
+			{
+				// Conditions met to transition into the trigger state.
+				if(i == currentState)
+					SetState(Body::BodyState::TRIGGER);
```
```suggestion
					SetState(BodyState::TRIGGER);
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -328,6 +683,80 @@ void Body::PauseAnimation()
+{
+	// Never ready for action if transitioning between states.
+	return anim.indicateReady ? stateReady : (!stateTransitionRequested ||
+														transitionState == Body::BodyState::TRIGGER);
```
```suggestion
		transitionState == BodyState::TRIGGER);
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -328,6 +683,80 @@ void Body::PauseAnimation()
+{
+	if(stateTransitionRequested)
+	{
+		integratedFrame = 0.0f;
```
```suggestion
		integratedFrame = 0.f;
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -328,6 +683,80 @@ void Body::PauseAnimation()
+		integratedFrame = 0.0f;
+		pause = 0;
+		// Current transition due to trigger.
+		bool triggerTransition = transitionState == Body::BodyState::TRIGGER;
+		Body::BodyState requestedTransitionState = triggerTransition ?
+									currentState : transitionState;
+		// Default to Flying sprite if requested sprite does not exist.
+		Body::BodyState trueTransitionState = sprites[requestedTransitionState].GetSprite() ?
+									requestedTransitionState : Body::BodyState::FLYING;
```
```suggestion
		bool triggerTransition = transitionState == BodyState::TRIGGER;
		BodyState requestedTransitionState = triggerTransition ?
			currentState : transitionState;
		// Default to Flying sprite if requested sprite does not exist.
		BodyState trueTransitionState = sprites[requestedTransitionState].GetSprite() ?
			requestedTransitionState : BodyState::FLYING;
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -328,6 +683,80 @@ void Body::PauseAnimation()
+		// Update animation parameters.
+		currentState = requestedTransitionState;
+		anim = *(transitionedState->GetExposedParameters());
+		if(anim.rampUpRate > 0.0)
```
```suggestion
		if(anim.rampUpRate > 0.)
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -328,6 +683,80 @@ void Body::PauseAnimation()
+		currentState = requestedTransitionState;
+		anim = *(transitionedState->GetExposedParameters());
+		if(anim.rampUpRate > 0.0)
+			frameRate = Body::MIN_FRAME_RATE;
```
```suggestion
			frameRate = MIN_FRAME_RATE;
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -328,6 +683,80 @@ void Body::PauseAnimation()
+
+
+
+
```
```suggestion
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -388,60 +819,150 @@ void Body::SetStep(int step) const
-		frame = fmod(frame, cycle);
-
-	if(!rewind)
+	int prevFrame = static_cast<int>(frame), nextFrame = -1;
```
I'm not sure if this is part of the official style guide, but we usually put every variable declaration on a separate line to improve code readability. Also I think that the explicit cast isn't necessary in this case.
```suggestion
	int prevFrame = frame;
	int nextFrame = -1;
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -388,60 +819,150 @@ void Body::SetStep(int step) const
-			frame = min(frame, lastFrame);
-		else if(frame >= frames)
+		// Handle any frameRate changes in the animation.
+		if(anim.rampUpRate != 0.0)
```
```suggestion
		if(anim.rampUpRate)
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -388,60 +819,150 @@ void Body::SetStep(int step) const
+		// Handle any frameRate changes in the animation.
+		if(anim.rampUpRate != 0.0)
+		{
+			if(frameRate >= Body::MIN_FRAME_RATE && frameRate <= anim.frameRate)
```
```suggestion
			if(frameRate >= MIN_FRAME_RATE && frameRate <= anim.frameRate)
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -388,60 +819,150 @@ void Body::SetStep(int step) const
+			frame = max(0.f, lastFrame * 2.f - frame);
+			stateReady = false;
 		}
+		nextFrame = static_cast<int>(frame);
+		if(nextFrame != prevFrame)
+			randomFrame = static_cast<int>(static_cast<float>(Random::Real()) * frames);
```
(see my previous comment)
```suggestion
		nextFrame = frame;
		if(nextFrame != prevFrame)
			randomFrame = Random::Real() * frames;
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -388,60 +819,150 @@ void Body::SetStep(int step) const
-		// be less than 0, clamp it to 0.
-		frame = max(0.f, lastFrame * 2.f - frame);
+		// Override any delay if the ship wants to jump, become disabled, or land.
+		bool ignoreDelay = transitionState == Body::BodyState::JUMPING ||
+							transitionState == Body::BodyState::DISABLED ||
+							transitionState == Body::BodyState::LANDING ||
+							transitionState == Body::BodyState::TRIGGER;
```
```suggestion
		bool ignoreDelay = transitionState == BodyState::JUMPING ||
			transitionState == BodyState::DISABLED ||
			transitionState == BodyState::LANDING ||
			transitionState == BodyState::TRIGGER;
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -388,60 +819,150 @@ void Body::SetStep(int step) const
+
+		if(delayed >= anim.transitionDelay || ignoreDelay)
+		{
+			if(anim.rampDownRate != 0.0)
```
```suggestion
			if(anim.rampDownRate)
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -388,60 +819,150 @@ void Body::SetStep(int step) const
+		{
+			if(anim.rampDownRate != 0.0)
+			{
+				if(frameRate >= Body::MIN_FRAME_RATE && frameRate <= anim.frameRate)
```
```suggestion
				if(frameRate >= MIN_FRAME_RATE && frameRate <= anim.frameRate)
```

TomGoodIdea commented on source/Body.cpp:
```diff
@@ -388,60 +819,150 @@ void Body::SetStep(int step) const
+				// Rewind the ongoing state's animation, then transition.
+				frame = max(0.f, rewindFrame * 2.f - frame);
+
+				if(frame == 0.f)
```
```suggestion
				if(!frame)
```
just my personal preference

TomGoodIdea commented on source/Body.h:
Are all these variables marked as mutables because you want to use them in const context, or just because you like mutables?

XessWaffle replied:
I marked them as mutable since the SetStep function was initially const. My modifications to SetStep lead to it being necessary to allow the mutable variables to change. In the spirit of not modifying already existing function definitions to the best of my ability, mutable was the workaround. But if it does not make that big of a difference I would prefer the SetStep function to not be const, and remove the mutable designator.

TomGoodIdea replied:
If it doesn't create more problems than solves, you can remove the const. IMO it doesn't make sense to have to set nearly every member as mutable for a single function.

TomGoodIdea commented on source/Body.h:
```diff
@@ -33,6 +35,14 @@ class Sprite;
+	static constexpr float MIN_FRAME_RATE = 2. / 60.;
+
+
+
```
two blank lines between blocks
```suggestion
```

TomGoodIdea commented on source/Body.h:
```diff
@@ -41,8 +51,10 @@ class Body {
+	bool HasSpriteFor(BodyState state) const;
 	// Access the underlying Sprite object.
-	const Sprite *GetSprite() const;
+	const Sprite *GetSprite(Body::BodyState state = Body::BodyState::CURRENT) const;
+	Body::BodyState GetState() const;
```
```suggestion
	const Sprite *GetSprite(BodyState state = BodyState::CURRENT) const;
	BodyState GetState() const;
```

TomGoodIdea commented on source/Body.h:
```diff
@@ -70,11 +82,15 @@ class Body {
-	void SaveSprite(DataWriter &out, const std::string &tag = "sprite") const;
+	// Sprite serialization. Return true if sprite is successfully loaded.
+	bool LoadSprite(const DataNode &node);
+	void LoadTriggerSprite(const DataNode &node, Body::BodyState state,
+							SpriteParameters::AnimationParameters params, int index);
```
```suggestion
	void LoadTriggerSprite(const DataNode &node, BodyState state,
		SpriteParameters::AnimationParameters params, int index);
```

TomGoodIdea commented on source/Body.h:
```diff
@@ -70,11 +82,15 @@ class Body {
+	void SaveSpriteParameters(DataWriter &out, SpriteParameters *state, int index) const;
 	// Set the sprite.
-	void SetSprite(const Sprite *sprite);
+	void SetSprite(const Sprite *sprite, Body::BodyState state = BodyState::FLYING);
+	void SetState(Body::BodyState state);
```
```suggestion
	void SetSprite(const Sprite *sprite, BodyState state = BodyState::FLYING);
	void SetState(BodyState state);
```

TomGoodIdea commented on source/Body.h:
```diff
@@ -99,48 +122,58 @@ class Body {
 	float zoom = 1.f;
-	float scale = 1.f;
+	mutable float scale = 1.f;
+	mutable bool debug = false;
```
What is this for?

XessWaffle replied:
Scale refers to the size of the sprite. Since the scale can be different across different sprites of the same ship (depending on design), I added it as a part of SpriteParameters. Every time a ship enters a new state, the SpriteParameters for that state is requested which causes the mutable variables to be refreshed according to the new SpriteParameters.

Debug was just something I had introduced to test an issue with the the trigger sprites and I forgot to remove it. I will remove it now.

TomGoodIdea commented on source/Body.h:
```diff
@@ -99,48 +122,58 @@ class Body {
 private:
 	// Animation parameters.
-	const Sprite *sprite = nullptr;
+	mutable SpriteParameters sprites[Body::BodyState::NUM_STATES] = {SpriteParameters(), SpriteParameters(),
+																SpriteParameters(), SpriteParameters(),
+																SpriteParameters(), SpriteParameters()};
```
Wouldn't it be just an empty array?
```suggestion
	mutable SpriteParameters sprites[BodyState::NUM_STATES] = {};
```

TomGoodIdea commented on source/Body.h:
```diff
@@ -99,48 +122,58 @@ class Body {
+	mutable SpriteParameters sprites[Body::BodyState::NUM_STATES] = {SpriteParameters(), SpriteParameters(),
+																SpriteParameters(), SpriteParameters(),
+																SpriteParameters(), SpriteParameters()};
+	mutable Body::BodyState currentState = Body::BodyState::FLYING,
+							transitionState = Body::BodyState::FLYING;
```
```suggestion
	mutable BodyState currentState = BodyState::FLYING;
	mutable BodyState transitionState = BodyState::FLYING;
```
(see one of my previous comments)

TomGoodIdea commented on source/Ship.cpp:
```diff
@@ -262,7 +267,7 @@ void Ship::Load(const DataNode &node)
 			continue;
 		}
-		if(key == "sprite")
+		if(key.find("sprite") != std::string::npos)
```
```suggestion
		if(key.find("sprite") != string::npos)
```

TomGoodIdea commented on source/Ship.cpp:
```diff
@@ -1447,6 +1452,13 @@ void Ship::SetGovernment(const Government *government)
 
+void Ship::SetIsPlayerFlagship(bool isFlagship)
+{
+	this->isPlayerFlagship = isFlagship;
```
```suggestion
	isPlayerFlagship = isFlagship;
```

TomGoodIdea commented on source/Ship.cpp:
```diff
@@ -2404,7 +2427,10 @@ bool Ship::IsReadyToJump(bool waitingIsReady) const
-	return true;
+	// For any ship that is not the player flagship,
+	// jumps should not be restricted by animation if they are not in the system.
+	return (GetState() == Body::BodyState::JUMPING && ReadyForAction()) ||
```
```suggestion
	return (GetState() == BodyState::JUMPING && ReadyForAction()) ||
```
Would this work? I mean, a `Ship` is a `Body`.

XessWaffle replied:
Seems to work!

TomGoodIdea commented on source/Ship.cpp:
```diff
@@ -2404,7 +2427,10 @@ bool Ship::IsReadyToJump(bool waitingIsReady) const
+	// For any ship that is not the player flagship,
+	// jumps should not be restricted by animation if they are not in the system.
+	return (GetState() == Body::BodyState::JUMPING && ReadyForAction()) ||
+			!(isPlayerFlagship || isInSystem);
```
```suggestion
		!(isPlayerFlagship || isInSystem);
```

TomGoodIdea commented on source/Ship.cpp:
```diff
@@ -3767,6 +3793,62 @@ const vector<weak_ptr<Ship>> &Ship::GetEscorts() const
+		if(targetSystem)
+		{
+			pair<JumpType, double> jumpUsed = navigation.GetCheapestJumpType(targetSystem);
+			bool jumpDrive = (jumpUsed.first == JumpType::JUMP_DRIVE) && outfit->TrueName() == "Jump Drive";
+			bool hyperDrive = (jumpUsed.first == JumpType::HYPERDRIVE) &&
+								(outfit->TrueName() == "Hyperdrive" || outfit->TrueName() == "Scram Drive");
```
```suggestion
				(outfit->TrueName() == "Hyperdrive" || outfit->TrueName() == "Scram Drive");
```
Also does this account for plugin outfits?

XessWaffle replied:
Good point, the only real check being done here is the type of jump, so maybe I don't need to define the actual drive being used. I will rewrite this to not require the type of drive.

TomGoodIdea commented on source/Ship.cpp:
```diff
@@ -3767,6 +3793,62 @@ const vector<weak_ptr<Ship>> &Ship::GetEscorts() const
+		for(unsigned i = 0; i < hardpoints.size(); ++i)
+		{
+			const Weapon *weapon = hardpoints[i].GetOutfit();
+			std::string weaponName = hardpoints[i].GetOutfit()->TrueName();
```
```suggestion
			string weaponName = hardpoints[i].GetOutfit()->TrueName();
```

TomGoodIdea commented on source/Ship.cpp:
```diff
@@ -4016,7 +4098,10 @@ void Ship::DoGeneration()
 				if(ship.HasDeployOrder())
+				{
 					DoRepair(ship.energy, energy, ship.attributes.Get("energy capacity"));
+					ship.SetState(Body::BodyState::LAUNCHING);
```
```suggestion
					ship.SetState(BodyState::LAUNCHING);
```
(see one of my previous comments)

TomGoodIdea commented on source/Ship.cpp:
```diff
@@ -4216,6 +4299,8 @@ void Ship::DoGeneration()
 				heat -= activeCooling * min(1., Heat());
 		}
 	}
+	else if(!HasSpriteFor(Body::BodyState::DISABLED))
```
```suggestion
	else if(!HasSpriteFor(BodyState::DISABLED))
```

TomGoodIdea commented on source/Ship.cpp:
```diff
@@ -4324,6 +4409,8 @@ bool Ship::DoHyperspaceLogic(vector<Visual> &visuals)
 	if(!hyperspaceSystem && !hyperspaceCount)
 		return false;
 
+	SetState(Body::BodyState::JUMPING);
```
```suggestion
	SetState(BodyState::JUMPING);
```

TomGoodIdea commented on source/Ship.cpp:
```diff
@@ -4467,6 +4554,8 @@ bool Ship::DoLandingLogic()
 	if(!landingPlanet && zoom >= 1.f)
 		return false;
 
+	static const double zoomTriggerStart = 0.03f;
```
```suggestion
	static const double zoomTriggerStart = .03f;
```

TomGoodIdea commented on source/Ship.cpp:
```diff
@@ -4481,12 +4570,15 @@ bool Ship::DoLandingLogic()
 	// just slowly refuel.
 	if(landingPlanet && zoom)
 	{
+		SetState(Body::BodyState::LANDING);
```
```suggestion
		SetState(BodyState::LANDING);
```

TomGoodIdea commented on source/Ship.cpp:
```diff
@@ -4508,11 +4600,23 @@ bool Ship::DoLandingLogic()
+		}
+		else if(zoom >= zoomTriggerStart)
+			ShowDefaultSprite(false);
+		SetState(Body::BodyState::LAUNCHING);
```
```suggestion
		SetState(BodyState::LAUNCHING);
```

TomGoodIdea commented on source/Ship.cpp:
```diff
@@ -4534,16 +4638,23 @@ void Ship::DoInitializeMovement()
 	// If you're disabled, you can't initiate landing or jumping.
 	if(isDisabled)
+	{
+		SetState(Body::BodyState::DISABLED);
```
```suggestion
		SetState(BodyState::DISABLED);
```

TomGoodIdea commented on source/Ship.cpp:
```diff
@@ -4534,16 +4638,23 @@ void Ship::DoInitializeMovement()
-		pair<JumpType, double> jumpUsed = navigation.GetCheapestJumpType(hyperspaceSystem);
-		isUsingJumpDrive = (jumpUsed.first == JumpType::JUMP_DRIVE);
-		hyperspaceFuelCost = jumpUsed.second;
+		SetState(Body::BodyState::JUMPING);
```
```suggestion
		SetState(BodyState::JUMPING);
```

TomGoodIdea commented on source/Ship.cpp:
```diff
@@ -4788,70 +4899,118 @@ void Ship::DoMovement(bool &isUsingAfterburner)
 			{
-				if(cloak && !attributes.Get("cloaked boarding"))
+				bool targetInRange = target->Position().Distance(Position()) < WEAPONS_RANGE_MULTIPLIER * weaponRange
+									|| !weaponRange;
```
```suggestion
					|| !weaponRange;
```

TomGoodIdea commented on source/Ship.cpp:
```diff
@@ -4788,70 +4899,118 @@ void Ship::DoMovement(bool &isUsingAfterburner)
-					// not to actually board, if they are cloaked, except if they have "cloaked boarding".
-					if(isYours)
-						Messages::Add("You cannot board a ship while cloaked.", Messages::Importance::High);
+					SetState(Body::BodyState::FIRING);
```
```suggestion
					SetState(BodyState::FIRING);
```

TomGoodIdea commented on source/Ship.cpp:
```diff
@@ -4788,70 +4899,118 @@ void Ship::DoMovement(bool &isUsingAfterburner)
 				}
 				else
-				{
+					SetState(Body::BodyState::FLYING);
```
```suggestion
					SetState(BodyState::FLYING);
```

TomGoodIdea commented on source/Ship.cpp:
```diff
@@ -4788,70 +4899,118 @@ void Ship::DoMovement(bool &isUsingAfterburner)
+					SetState(Body::BodyState::FLYING);
+			}
+			else if(hasPrimary)
+				SetState(Body::BodyState::FIRING);
```
```suggestion
				SetState(BodyState::FIRING);
```

TomGoodIdea commented on source/Ship.cpp:
```diff
@@ -4788,70 +4899,118 @@ void Ship::DoMovement(bool &isUsingAfterburner)
+			// Fighter/Drone is boarding/landing.
+			if(isBoarding && canBeCarriedNow)
+			{
+				SetState(Body::BodyState::LANDING);
```
```suggestion
				SetState(BodyState::LANDING);
```

TomGoodIdea commented on source/Ship.cpp:
```diff
@@ -4788,70 +4899,118 @@ void Ship::DoMovement(bool &isUsingAfterburner)
+				if(target && !isDisabled)
+				{
+					bool targetInRange = target->Position().Distance(Position()) < WEAPONS_RANGE_MULTIPLIER * weaponRange
+										|| weaponRange == 0.0;
```
```suggestion
						|| !weaponRange;
```

TomGoodIdea commented on source/Ship.cpp:
```diff
@@ -4788,70 +4899,118 @@ void Ship::DoMovement(bool &isUsingAfterburner)
+										|| weaponRange == 0.0;
+					// If in range, or the weapon range hasn't been calculated yet.
+					if(targetInRange)
+						SetState(Body::BodyState::FIRING);
```
```suggestion
						SetState(BodyState::FIRING);
```

TomGoodIdea commented on source/Ship.cpp:
```diff
@@ -4788,70 +4899,118 @@ void Ship::DoMovement(bool &isUsingAfterburner)
+				}
+				// If the ship has no target but is still flying around and doesn't want to jump.
+				else
+					SetState(Body::BodyState::FLYING);
```
```suggestion
					SetState(BodyState::FLYING);
```

TomGoodIdea commented on source/Ship.cpp:
```diff
@@ -4788,70 +4899,118 @@ void Ship::DoMovement(bool &isUsingAfterburner)
+					SetState(Body::BodyState::FLYING);
+			}
+			else if(hasPrimary)
+				SetState(Body::BodyState::FIRING);
```
```suggestion
				SetState(BodyState::FIRING);
```

TomGoodIdea commented on source/Ship.h:
```diff
@@ -589,13 +591,15 @@ class Ship : public Body, public std::enable_shared_from_this<Ship> {
 	bool neverDisabled = false;
 	bool isCapturable = true;
 	bool isInvisible = false;
+	bool isPlayerFlagship = false;
```
If we have this variable, I think it could be used in many places in AI.cpp (not in this PR, it's already big).

XessWaffle replied:
Could be a future PR

TomGoodIdea commented on source/SpriteParameters.cpp:
```diff
@@ -0,0 +1,158 @@
+
+#include "Sprite.h"
+
+
```
```suggestion
using namespace std;


```
`std::get<>`s can probably stay in the current form, but I'd like to hear others' opinions

XessWaffle replied:
`get<>` does feel a little ambiguous, I think for clarity's sake we keep it that way, but I don't really have a preference either way.

RisingLeaf replied:
keep it imo

TomGoodIdea commented on source/SpriteParameters.cpp:
```diff
@@ -0,0 +1,158 @@
+SpriteParameters::SpriteParameters(const Sprite *sprite)
+{
+	AnimationParameters initDefault;
+	auto tuple = SpriteParameters::SpriteDetails{sprite, initDefault, {}};
```
Possible future "tuple" variable name ambiguity problem.

TomGoodIdea commented on source/SpriteParameters.cpp:
```diff
@@ -0,0 +1,158 @@
+{
+	AnimationParameters initDefault;
+	auto tuple = SpriteParameters::SpriteDetails{sprite, initDefault, {}};
+	sprites.insert(std::pair<int, SpriteParameters::SpriteDetails>(0, tuple));
```
```suggestion
	sprites.insert(pair<int, SpriteParameters::SpriteDetails>(0, tuple));
```

TomGoodIdea commented on source/SpriteParameters.cpp:
```diff
@@ -0,0 +1,158 @@
+
+
+void SpriteParameters::SetSprite(int index, const Sprite *sprite, SpriteParameters::AnimationParameters data,
+		ConditionSet triggerConditions)
```
```suggestion
	ConditionSet triggerConditions)
```

TomGoodIdea commented on source/SpriteParameters.cpp:
```diff
@@ -0,0 +1,158 @@
+		ConditionSet triggerConditions)
+{
+	auto tuple = SpriteParameters::SpriteDetails{sprite, data, triggerConditions};
+	sprites.insert(std::pair<int, SpriteParameters::SpriteDetails>(index, tuple));
```
```suggestion
	sprites.insert(pair<int, SpriteParameters::SpriteDetails>(index, tuple));
```

TomGoodIdea commented on source/SpriteParameters.cpp:
```diff
@@ -0,0 +1,158 @@
+	if(it != sprites.end())
+		return std::get<0>(it->second);
+
+	return NULL;
```
```suggestion
	return nullptr;
```

TomGoodIdea commented on source/SpriteParameters.cpp:
```diff
@@ -0,0 +1,158 @@
+
+ConditionSet SpriteParameters::GetConditions(int index) const
+{
+	static ConditionSet empty;
+	if(index < 0)
+		return std::get<2>(exposedDetails);
+
+	auto it = sprites.find(index);
+	if(it != sprites.end())
+		return std::get<2>(it->second);
+
+	return empty;
```
```suggestion
	if(index < 0)
		return std::get<2>(exposedDetails);

	auto it = sprites.find(index);
	if(it != sprites.end())
		return std::get<2>(it->second);

	return {};
```

TomGoodIdea commented on source/SpriteParameters.cpp:
```diff
@@ -0,0 +1,158 @@
+
+SpriteParameters::AnimationParameters SpriteParameters::GetParameters(int index) const
+{
+	SpriteParameters::AnimationParameters def;
+	if(index < 0)
+		return std::get<1>(exposedDetails);
+
+	auto it = sprites.find(index);
+	if(it != sprites.end())
+		return std::get<1>(it->second);
+
+	return def;
```
```suggestion
	if(index < 0)
		return std::get<1>(exposedDetails);

	auto it = sprites.find(index);
	if(it != sprites.end())
		return std::get<1>(it->second);

	return {};
```

TomGoodIdea commented on source/SpriteParameters.cpp:
```diff
@@ -0,0 +1,158 @@
+
+
+
+const int SpriteParameters::GetExposedID() const
```
No need for const fundamental returns.
```suggestion
int SpriteParameters::GetExposedID() const
```

TomGoodIdea commented on source/SpriteParameters.cpp:
```diff
@@ -0,0 +1,158 @@
+
+
+
+const int SpriteParameters::RequestTriggerUpdate(ConditionsStore &store)
```
(as above)
```suggestion
int SpriteParameters::RequestTriggerUpdate(ConditionsStore &store)
```

TomGoodIdea commented on source/SpriteParameters.h:
```diff
@@ -0,0 +1,121 @@
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef SPRITE_PARAMETERS_H
+#define SPRITE_PARAMETERS_H
```
```suggestion
#ifndef SPRITE_PARAMETERS_H_
#define SPRITE_PARAMETERS_H_
```

TomGoodIdea commented on source/SpriteParameters.h:
```diff
@@ -0,0 +1,121 @@
+	};
+
+
+
```
```suggestion
```

TomGoodIdea commented on source/SpriteParameters.h:
```diff
@@ -0,0 +1,121 @@
+	typedef std::map<int, SpriteParameters::SpriteDetails> SpriteMap;
+
+
+
```
```suggestion
```

TomGoodIdea commented on source/SpriteParameters.h:
```diff
@@ -0,0 +1,121 @@
+	ConditionSet GetConditions(int index = -1) const;
+	AnimationParameters GetParameters(int index = -1) const;
+	// Used to verify trigger transitions.
+	const int GetExposedID() const;
+	// Check all conditions and return if one is true.
+	const int RequestTriggerUpdate(ConditionsStore &store);
```
```suggestion
	int GetExposedID() const;
	// Check all conditions and return if one is true.
	int RequestTriggerUpdate(ConditionsStore &store);
```

TomGoodIdea commented on source/SpriteParameters.h:
```diff
@@ -0,0 +1,121 @@
+
+public:
+	// ID For default sprite.
+	const static int DEFAULT = 0;
```
Do we really need this 0 const?

XessWaffle replied:
I made it for clarity, but I don't think it is necessary.

TomGoodIdea commented on source/SpriteParameters.h:
```diff
@@ -0,0 +1,121 @@
+
+private:
+	// Sprites to be animated.
+	SpriteParameters::SpriteMap sprites;
+	SpriteParameters::SpriteDetails exposedDetails, defaultDetails;
+	// Animation parameters exposed to Body.
+	SpriteParameters::AnimationParameters exposed;
+	int exposedIndex = 0, requestedIndex = 0;
```
```suggestion
	SpriteMap sprites;
	SpriteDetails exposedDetails;
	SpriteDetails defaultDetails;
	// Animation parameters exposed to Body.
	AnimationParameters exposed;
	int exposedIndex = 0;
	int requestedIndex = 0;
```

TomGoodIdea reviewed (commented):
You'll need to make a PR to [the wiki repo](https://github.com/endless-sky/endless-sky-wiki) with documentation on this feature.

tibetiroka commented:
I haven't been able to do a proper review yet, but this looks really cool. One bug I noticed is that you can't jump while firing weapons, even if they have no firing force.

XessWaffle commented on source/Body.cpp:
```diff
@@ -205,90 +251,399 @@ const Government *Body::GetGovernment() const
+		}
+		else if(child.Token(0) == "transition rewind")
+		{
+			spriteAnimationParameters.transitionFinish = false;
+			spriteAnimationParameters.transitionRewind = true;
```
They are not necessarily mutually exclusive, there is a case where both of them are false where the sprite animation behaves differently. However, now that you point this out, I think this can be better designed i.e. have one input `transition type` and have a defined enumeration:

- `transition type 0`: Transition immediately into the next state
- `transition type 1`: Transition after finishing the current animation; same as `transitionFinish`
- `transition type 2`: Transition after rewinding the current animation; same as `transitionRewind`

This could also make it easier for newer transition types to be added in the future.

XessWaffle commented:
Thanks for taking a look @TomGoodIdea! I hopefully addressed most of the concerns.


> I love this. I'd very much welcome the ability to have deploy animations and separate flying/docked fighter sprites. I'm imagining something similar to how an MV-22 Osprey can fold up its wings to pack into a tighter space. That would allow for a lot more variation in fighter design, and this pr as a whole I think really opens up what people can do. Variable geometry spacecraft are absolutely something I'd love to see in this game and this is a huge step towards making that happen.

@Hyugat, most of this should already be in the scope of the PR! Fighters can have separate launch and land (docking) animations when leaving or being recalled from the mothership bay. Having a separate deploy state for the mothership is something I am looking into creating. This would allow for something like bay doors opening in order to recall/deploy fighters. Not sure yet whether that will make it into this PR though.


> I haven't been able to do a proper review yet, but this looks really cool. One bug I noticed is that you can't jump while firing weapons, even if they have no firing force.

@tibetiroka good catch, by current design logic, a Ship cannot be in two animation states at once, this would lead to all sorts of problems in determining which image to play next in the animation. JUMPING and FIRING are two separate states and therefore the ship shouldn't enter the JUMPING animation while it is doing the FIRING animation. It would need to transition out of the FIRING animation state in order to do so. I will have to look into how to reconcile this for the edge case where we are firing weapons and are ready to jump.

tibetiroka commented:
I looked at the wiki PR for the data format, and I have some comments:
- Are `ramp up` and `random` essential to this PR? I feel like they could be split off and merged separately. (Maybe also `reverse`, though that's used here.)
- `"transition type"` should take textual parameters (`immediate`, `finish`, `rewind`), and convert them to numbers internally.
- `indicate` and `"indicate percentage"` could be merged to a single `indicate` option with an optional parameter. I also find it weird that it takes a percentage value instead of a frame count like the other options.

RisingLeaf commented on source/SpriteParameters.h:
```diff
@@ -0,0 +1,121 @@
+public:
+
+	// Supported animation transition types from one state to another.
+	enum TransitionType{IMMEDIATE, FINISH, REWIND, NUM_TRANSITIONS};
+
```
```suggestion
	enum TransitionType {
		IMMEDIATE,
		FINISH,
		REWIND,
		NUM_TRANSITIONS
	};

```

RisingLeaf commented on source/SpriteParameters.h:
```diff
@@ -0,0 +1,121 @@
+
+public:
+	typedef std::tuple<const Sprite*, SpriteParameters::AnimationParameters, ConditionSet> SpriteDetails;
+	typedef std::map<int, SpriteParameters::SpriteDetails> SpriteMap;
```
this really does not need a typedef

XessWaffle replied:
The typedef was simply to keep the length of the variable declaration down in the class otherwise in the cpp file we would be seeing a lot of large declaration types. I would like to keep it this way for clarity, are there any alternatives to this?

tibetiroka replied:
You should probably be using [using](https://en.cppreference.com/w/cpp/language/type_alias).

TomGoodIdea replied:
if you remove the unnecessary prefixes, `SpriteParameters::` (you don't need these as you are inside this class) and `std::` (you don't need them in the `.cpp` file), they'll become much shorter

XessWaffle replied:
I have shortened up the names in the latest commit, take a look and let me know if the solution is appropriate!

RisingLeaf commented on source/SpriteParameters.cpp:
```diff
@@ -0,0 +1,157 @@
+
+SpriteParameters::SpriteParameters(const Sprite *sprite)
+{
+	AnimationParameters initDefault;
+	auto spriteDetails = SpriteParameters::SpriteDetails{sprite, initDefault, {}};
+	sprites.insert(pair<int, SpriteParameters::SpriteDetails>(0, spriteDetails));
+	exposed = initDefault;
```
```suggestion
	auto spriteDetails = SpriteParameters::SpriteDetails{sprite, {}, {}};
	sprites.insert(pair<int, SpriteParameters::SpriteDetails>(0, spriteDetails));
```

XessWaffle replied:
Good catch, I should look through the code and see if I can find any other places where I can use `{}`

XessWaffle commented:
> I looked at the wiki PR for the data format, and I have some comments:
> 
> * Are `ramp up` and `random` essential to this PR? I feel like they could be split off and merged separately. (Maybe also `reverse`, though that's used here.)
These are not essential to this PR, they were added as part of a request in one of the earlier closed PRs. But at this point separating them out into a different PR would be tougher as it would require me to touch the logic of the SetStep function, something which I have already modified quite a bit over the history of this change. Since the changes to SetStep are a core part of this PR, I would prefer not to touch them for now unless they cause animation breaking bugs.

That being said, I will take a look, and if it does not mess with SetStep too much, I will separate them out into a different PR, one which will be required to merge along with this one.

> * `"transition type"` should take textual parameters (`immediate`, `finish`, `rewind`), and convert them to numbers internally.
> * `indicate` and `"indicate percentage"` could be merged to a single `indicate` option with an optional parameter. I also find it weird that it takes a percentage value instead of a frame count like the other options.

I agree with both of these points, let me make the changes.

tibetiroka commented:
Requesting Saugia for the sprite syntax.

TomGoodIdea commented on source/SpriteParameters.h:
```diff
@@ -0,0 +1,126 @@
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef SPRITE_PARAMETERS_H_
+#define SPRITE_PARAMETERS_H_
```
we switched to `#pragma once` a while ago
```suggestion
#pragma once
```

TomGoodIdea commented on source/SpriteParameters.h:
```diff
@@ -0,0 +1,126 @@
+
+class Sprite;
+
+using namespace std;
```
we don't do this in headers
```suggestion
```

TomGoodIdea commented on source/SpriteParameters.h:
```diff
@@ -0,0 +1,126 @@
+		Point center;
+	};
+
+	using SpriteDetails = tuple<const Sprite*, AnimationParameters, ConditionSet>;
+	using SpriteMap = map<int, SpriteDetails>;
```
```suggestion
	using SpriteDetails = std::tuple<const Sprite *, AnimationParameters, ConditionSet>;
	using SpriteMap = std::map<int, SpriteDetails>;
```

TomGoodIdea commented on source/SpriteParameters.h:
```diff
@@ -0,0 +1,126 @@
+	int exposedIndex = 0;
+	int requestedIndex = 0;
+};
+
+
+
+#endif
```
```suggestion

```

TomGoodIdea commented:
Also @XessWaffle please resolve conflicts

TomGoodIdea commented:
Could you update the PR description to match the current data syntax?
